lib/utils/resolveSilent.mjs

```diff
@@ -0,0 +1,53 @@
+import fs from 'node:fs';
+import path from 'node:path';
+
+import { fileURLToPath, pathToFileURL } from 'node:url';
+
+// TODO: use pure ESM `import-meta-resolve` package in next major version
+import { moduleResolve } from '@dual-bundle/import-meta-resolve';
+import resolveFrom from 'resolve-from';
+
+/**
+ * TODO: These suffixes are here for commonjs compatibility reason, we should remove these codes after migrating to pure ESM, because in ESM accurate paths are required
+ */
+
+const pathSuffixes = ['', '.js', '.json', `${path.sep}index.js`, `${path.sep}index.json`];
+
+const specifierSuffixes = ['', '.js', '.json', '/index.js', '/index.json'];
+
+/**
+ * @param {string} parent
+ * @param {string} lookup
+ * @return {string | undefined}
+ */
+export default function resolveSilent(parent, lookup) {
+	if (path.isAbsolute(lookup)) {
+		for (const suffix of pathSuffixes) {
+			const filename = lookup + suffix;
+
+			if (fs.existsSync(filename)) {
+				return filename;
+			}
+		}
+
+		return;
+	}
+
+	const base = pathToFileURL(path.resolve(parent, 'noop.js'));
+
+	for (const suffix of specifierSuffixes) {
+		try {
+			return fileURLToPath(moduleResolve(lookup + suffix, base));
+		} catch {
+			//
+		}
+	}
+
+	/**
+	 * Yarn P'n'P does not support pure ESM well by default, this is only a workaround for it
+	 * @see https://github.com/wooorm/import-meta-resolve/issues/23
+	 *
+	 * TODO: this workaround is still necessary before native `import.meta.resolve` replacement
```

Comment:
[ask] Thanks for the TODO comment. I'm reading the Yarn doc (https://yarnpkg.com/getting-started/qa#does-yarn-support-esm) from your shared issue. So let me ask.

The Yarn doc says:

> Loaders aren't perfect yet, and until they are ESM-only packages cannot be recommended, but there's a way forward and as we work together we'll get there. We just have to be careful not to push people aside as we run towards our goal.

Does this mean that Yarn will resolve this ESM issue in the future, i.e. will this workaround no longer be necessary? If so, is there any GitHub issue we should track?

Reply:
When we migrate to native `import.meta.resolve`, yarn P'n'P will be supported as well.

The issue here is that `import-meta-resolve` does not support `--loader`.

Reply:
I see. Thanks for the answer 👍🏼